csrc/val_graph_visitor.h

```diff
@@ -120,139 +120,80 @@ class ValGraphStmtSort : public ValGraphVisitor {
   ValGroups sorted_vals_;
 };
 
-enum class Direction { Forward, Backward, Undefined };
-
-std::ostream& operator<<(std::ostream&, const Direction);
+class ValGraphDefinitions {
+  const ValGraph& graph_;
 
-using ExprPath = std::vector<std::pair<ExprGroup, Direction>>;
+ public:
+  ValGraphDefinitions(const ValGraph& graph) : graph_(graph) {}
+  decltype(auto) operator()(const ValGroup& val_group) const {
```

Comment:
Why doesn't just `auto` instead of `decltype(auto)` work here?

Reply:
`decltype(auto)` faithfully preserves the declare type, `auto` does not. For example:

```C++
int x;

const int& a() {
    return x;
}

auto b() { // returns int
    return a();
}

decltype(auto) c() { // return const int&
    return a();
}

int main() {
    auto d = a();  // int
    decltype(auto) e = a();  // const int&
}
```

Reply:
TIL. Thanks